airflow/contrib/hooks/gcp_vision_hook.py

```diff
@@ -0,0 +1,245 @@
+        self.log.info('Retrieving Product: %s', name)
+        response = self._handle_request(lambda **kwargs: client.get_product(**kwargs), name=name, retry=retry,
+                                        timeout=timeout, metadata=metadata)
+        self.log.info('Product retrieved.')
+        self.log.debug('Product retrieved:\n%s', response)
+        return MessageToJson(response)
+
+    @GoogleCloudBaseHook.fallback_to_default_project_id
+    def update_product(self, product, location=None, product_id=None, update_mask=None, project_id=None,
+                       retry=None, timeout=None, metadata=None):
+        """
+        For the documentation see:
+        :py:class:`~airflow.contrib.operators.gcp_vision_operator.CloudVisionProductUpdateOperator`
+        """
+        client = self.get_client()
+        product = self._determine_name(product, product_id, location, project_id)
+        self.log.info('Updating ProductSet: %s', product.name)
+        response = self._handle_request(lambda **kwargs: client.update_product(**kwargs), product=product,
+                                        update_mask=update_mask, retry=retry, timeout=timeout,
+                                        metadata=metadata)
+        self.log.info('Product updated: %s', response.name if response else '')
+        self.log.debug('Product updated:\n%s', response)
+        return MessageToJson(response)
+
+    @GoogleCloudBaseHook.fallback_to_default_project_id
+    def delete_product(self, location, product_id, project_id=None, retry=None, timeout=None, metadata=None):
+        """
+        For the documentation see:
+        :py:class:`~airflow.contrib.operators.gcp_vision_operator.CloudVisionProductDeleteOperator`
+        """
+        client = self.get_client()
+        name = PRODUCT_NAME_TEMPLATE.format(project_id, location, product_id)
+        self.log.info('Deleting ProductSet: %s', name)
+        response = self._handle_request(lambda **kwargs: client.delete_product(**kwargs), name=name,
+                                        retry=retry, timeout=timeout, metadata=metadata)
+        self.log.info('Product with the name [%s] deleted:', name)
+        return response
+
+    def _handle_request(self, fun, **kwargs):
+        try:
+            return fun(**kwargs)
+        except GoogleAPICallError as e:
+            if isinstance(e, AlreadyExists):
+                raise e
+            else:
+                self.log.error('The request failed:\n%s', str(e))
+                raise AirflowException(e)
+        except RetryError as e:
+            self.log.error('The request failed due to a retryable error and retry '
+                           'attempts failed.')
+            raise AirflowException(e)
+        except ValueError as e:
+            self.log.error('The request failed, the parameters are invalid.')
+            raise AirflowException(e)
+
+    @staticmethod
+    def _get_autogenerated_id(response):
```

Comment:
If I understand it correctly, you can shorten it to
```python
name = response.get('name')
if '/' not in name:
            raise AirflowException('Unable to get name from response... [{}]\n{}'.format(response, e))
return name.rsplit('/', 1)[1]
```

Reply:
After having added all the necessary validations the difference is minimal, 8 lines versus 10:
```
try:
            name = response.name
        except AttributeError as e:
            raise AirflowException('Unable to get name from response... [{}]\n{}'.format(response, e))
        product_set_id = name.rsplit('/', 1)[1]
        if '/' not in name or not product_set_id:
            raise AirflowException('Unable to get id from name... [{}]'.format(name))
        return product_set_id
```

But I'll change it.

Reply:
Ok, after adding some more tests I shortened it to 7 lines ;)
```
@staticmethod
    def _get_autogenerated_id(response):
        try:
            name = response.name
        except AttributeError as e:
            raise AirflowException('Unable to get name from response... [{}]\n{}'.format(response, e))
        if '/' not in name:
            raise AirflowException('Unable to get id from name... [{}]'.format(name))
        return name.rsplit('/', 1)[1]
```